client/js/spotify-gdpr.js

```diff
@@ -0,0 +1,198 @@
+// @ts-check
+/* global c3 */
+/* global d3 */
+
+const songs = [
+  {id: 'song1'},
+  {id: 'song2'},
+  {id: 'song3'},
+  {id: 'song4'},
+  {id: 'song5'},
+  {id: 'song6'},
+  {id: 'song7'},
+];
+
+const latest = [
+  'song3',
+  'song2',
+  'song5',
+  'song7',
+  'song1',
+  'song10',
+  'song11',
+  'song12',
+  'song13',
+  'song14',
+  'song15',
+  'song16',
+  'song17',
+  'song18',
+  'song19',
+  'song20',
+  'song21',
+  'song22',
+  'song23',
+  'song24',
+  'song25',
+  'song26',
+  'song27',
+  'song28',
+  'song29',
+];
+
+/**
+ * @typedef {object} MoveChange
+ * @property {'move'} type
+ * @property {number} from
+ * @property {number} to
+ */
+
+/**
+ * @typedef {object} ReplaceChange
+ * @property {'replace'} type
+ * @property {string} oldItem
+ * @property {string} newItem
+ */
+
+/**
+ * @typedef {MoveChange | ReplaceChange} Change
+ */
+
+/**
+ * @type {Change[][]}
+ */
+const changes = [
+  [{type: 'move', from: 2, to: 4}],
+  [{type: 'replace', oldItem: 'song4', newItem: 'song7'}],
+  [{type: 'move', from: 3, to: 4}],
+  [{type: 'move', from: 5, to: 4}],
+  [{type: 'move', from: 3, to: 4}],
+  [],
+  [],
+  [{type: 'replace', oldItem: 'song30', newItem: 'song23'}],
+  [{type: 'move', from: 6, to: 5}],
+  [{type: 'move', from: 7, to: 6}],
+  [{type: 'move', from: 9, to: 10}],
+  [{type: 'move', from: 20, to: 21}],
+  [],
+  [{type: 'move', from: 20, to: 21}],
+  [{type: 'move', from: 12, to: 13}],
+  [],
+  [],
+  [],
+];
+
+
+let generation = 1;
+const x = [new Date()];
+const current = latest;
+const indices = new Map(current.map((id, idx) => [id, idx]));
+const indicesHistory = new Map(current.map((id, idx) => [id, [idx]]));
+
+for (const changeSet of changes.reverse()) {
+  for (const change of changeSet.reverse()) {
+    switch (change.type) {
+    case 'move': {
+      const {from, to} = change;
+
+      const a = current[from];
+      const b = current[to];
+      current[from] = b;
+      current[to] = a;
+
+      indices.set(a, to);
+      indices.set(b, from);
+      break;
+    }
+    case 'replace': {
+      const {oldItem, newItem} = change;
+
+      const idx = indices.get(newItem);
+      current[idx] = oldItem;
+
+      indices.set(newItem, null);
+      indices.set(oldItem, idx);
+      break;
+    }
```

Comment:
could you add some comments? not really sure what's going on in chunks of code like this.

Reply:
you're right, i apologise